SSA/Experimental/Bits/Fast/BitStream.lean

```diff
@@ -320,12 +320,54 @@ theorem neg_neg : a = - - a := by
     a.neg.negAux i = ⟨a i, (a.negAux i).2⟩ := by
     induction' i with i ih
     · simp [neg, negAux]
-    · simp [negAux, neg, ih]
+    · simp [neg, negAux, ih, Bool.xor_not_xor, Bool.not_xor_and_self, -Bool.not_bne',-Bool.not_bne]
   simp [Neg.neg, neg, neg_lemma]
 
--- TODO: This sorry is difficult, and will be proven in a later Pull Request.
-theorem ofBitVec_sub : ofBitVec (x - y) ≈ʷ (ofBitVec x) - (ofBitVec y)  := by
-  sorry
+/--
+ Will subAux overflow/carry when computing a - b?
+-/
+def subCarries? (a b : BitStream) (i : Nat) : Bool :=
+  let carry : Bool := match i with
+  | 0 => false
+  | i + 1 => a.subCarries? b i
+  (!a i && b i || !xor (a i) (b i) && carry)
+
+/--
+  For any i : ℕ, either the ith bit of -b will not overflow,
+  or the ith bit of (a + -b) will not overflow.
+-/
+theorem neg_or_add (i : Nat) :
+    (b.negAux i).2 = false ∨ (a.addAux b.neg i).2 = false := by
+  induction' i with i ih
+  <;> simp only [negAux, addAux, BitVec.adcb, neg]
+  · cases b 0
+    <;> simp
+  · cases' ih with l l
+    <;> cases b (i + 1)
+    <;> cases a (i + 1)
+    <;> simp [l]
+
+/--
+  Whether a - b will overflow is equivalent to -b overflows = (a + - b) overflows.
+-/
+theorem subCarries?_correct (i : Nat) :
+    a.subCarries? b i = ((b.negAux i).2 == (a.addAux b.neg i).2) := by
+  induction' i with i ih
+  · simp [subCarries?, negAux, addAux, BitVec.adcb, neg]
+  · by_cases a1 : a (i + 1)
+    <;> by_cases b1 : b (i + 1)
+    <;> cases' @neg_or_add a b i with h h
+    <;> simp [h, subCarries?, ih, negAux, addAux, BitVec.adcb, neg, a1, b1]
+
+theorem subAux_inductive_lemma (i : Nat) :
+    a.subAux b i = ⟨(a.addAux b.neg i).1, subCarries? a b i⟩ := by
+  induction' i with i ih
```

Comment:
We generally prefer rather mechanical names, based on a theorems type, over names based on intended usage.
This could, e.g., be called `subAux_eq_addAux_neg`